### PR TITLE
Cleanup CodeFactor issues with empty strings

### PR DESCRIPTION
**Pull request by iSazonov**
## PR Summary

Please fast review to avoid merge conflicts because of many files changed.

Please review commit by commit - every commit contains only single kind of change so you can do very fast review.

This corrects about 1% of the CodeFactor issues.

The PR replace "" with string.Empty.

All changes is made in VS Code with Regex patterns - one pattern by commit:
```
'= "";' -> '= string.Empty;'

'return "";' -> 'return string.Empty;'

': "";' -> ': string.Empty;'

'\?\? "";' -> '?? string.Empty;'

'\(""\)' -> '(string.Empty)'

' ""\)' -> ' string.Empty)'

'\("",' -> '(string.Empty,'

', "",' -> ', string.Empty,'

',""\)' -> ', string.Empty)'

'\? "" :' -> '? string.Empty :'

'\?\? ""' -> '?? string.Empty'

' = "",' -> ' = string.Empty,'

': "",' -> ': string.Empty,'

'@""\)' -> 'string.Empty)'
```

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Review by markekraus — APPROVED** [review]
LGTM assuming tests pass

**Comment by iSazonov** [comment]
@markekraus Many thanks for fast review!

**Comment by lzybkr** [comment]
I don't see the point. Why not just disable the warning and skip the code churn?

**Comment by iSazonov** [comment]
It is _readability issue_. If we use advanced editor with code highlighting we haven't problems to see "". If we use a simple editor the string.Empty visibility is much better.
Also see my https://github.com/PowerShell/PowerShell/pull/6949#issuecomment-393030657

File: src/System.Management.Automation/DscSupport/CimDSCParser.cs [hunk 1]
```diff
@@ -2820,17 +2820,17 @@ private static string MapAttributesToMof(string[] enumNames, IEnumerable<object>
...
                     if (dscProperty.Key)
                     {
-                        sb.AppendFormat(CultureInfo.InvariantCulture, "{0}key", needComma ? ", " : "");
+                        sb.AppendFormat(CultureInfo.InvariantCulture, "{0}key", needComma ? ", " : string.Empty);
                         needComma = true;
                     }
                     if (dscProperty.Mandatory)
                     {
-                        sb.AppendFormat(CultureInfo.InvariantCulture, "{0}required", needComma ? ", " : "");
+                        sb.AppendFormat(CultureInfo.InvariantCulture, "{0}required", needComma ? ", " : string.Empty);
```
Comment (BrucePay) [inline src/System.Management.Automation/DscSupport/CimDSCParser.cs]: Could migrate all of the `", "` strings into a constant.

File: src/System.Management.Automation/cimSupport/cmdletization/xml/cmdlets-over-objects.xmlSerializer.autogen.cs [hunk 2]
```diff
@@ -35,7 +35,7 @@ public void Write51_ClassMetadata(object o)
             WriteStartDocument();
             if (o == null)
             {
-                WriteNullTagLiteral(@"ClassMetadata", @"");
+                WriteNullTagLiteral(@"ClassMetadata", string.Empty);
```
Comment (BrucePay) [inline src/System.Management.Automation/cimSupport/cmdletization/xml/cmdlets-over-objects.xmlSerializer.autogen.cs]: If this is an autogen'ed file should you be changing it or the generator?

**Review by BrucePay — APPROVED** [review]
Overall the results look good. All the changes seem fine.

File: src/System.Management.Automation/DscSupport/CimDSCParser.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/DscSupport/CimDSCParser.cs]: Constant or static?

File: src/System.Management.Automation/cimSupport/cmdletization/xml/cmdlets-over-objects.xmlSerializer.autogen.cs (hunk 2 above)
Comment (iSazonov) [reply src/System.Management.Automation/cimSupport/cmdletization/xml/cmdlets-over-objects.xmlSerializer.autogen.cs]: The files is not regenerated and already was changed manually. Seems we should rename the files and remove "autogen" from file names.

File: src/System.Management.Automation/DscSupport/CimDSCParser.cs (hunk 1 above)
Comment (BrucePay) [reply src/System.Management.Automation/DscSupport/CimDSCParser.cs]: Is there any reason it shouldn't be constant?

File: src/System.Management.Automation/cimSupport/cmdletization/xml/cmdlets-over-objects.xmlSerializer.autogen.cs (hunk 2 above)
Comment (BrucePay) [reply src/System.Management.Automation/cimSupport/cmdletization/xml/cmdlets-over-objects.xmlSerializer.autogen.cs]: And remove the comments saying "this is an autogen'ed  file, don't edit"

File: src/System.Management.Automation/DscSupport/CimDSCParser.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/DscSupport/CimDSCParser.cs]: We already have a lot of static variables for such strings.
I found 4487 hits in 581 file for ", ". So static may be useful.

File: src/System.Management.Automation/cimSupport/cmdletization/xml/cmdlets-over-objects.xmlSerializer.autogen.cs (hunk 2 above)
Comment (iSazonov) [reply src/System.Management.Automation/cimSupport/cmdletization/xml/cmdlets-over-objects.xmlSerializer.autogen.cs]: Tracking issue #6974

**Comment by iSazonov** [comment]
@SteveL-MSFT Can we merge?

**Comment by dantraMSFT** [comment]
@iSazonov What does indicate as the reasoning for this change?  Is this a readability or programmatic warning?

My understanding of "" is that it is interned and you will have one instance per assembly.  In all other regards, String.Empty is synonymous with "".

File: src/System.Management.Automation/engine/parser/token.cs [hunk 3]
```diff
@@ -1045,25 +1045,25 @@ public static class TokenTraits
...
+            /*    Reserved slot 11  */ string.Empty,
+            /*    Reserved slot 12  */ string.Empty,
+            /*    Reserved slot 13  */ string.Empty,
+            /*    Reserved slot 14  */ string.Empty,
+            /*    Reserved slot 15  */ string.Empty,
+            /*    Reserved slot 16  */ string.Empty,
+            /*    Reserved slot 17  */ string.Empty,
+            /*    Reserved slot 18  */ string.Empty,
+            /*    Reserved slot 19  */ string.Empty,
+            /*    Reserved slot 20  */ string.Empty,
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/parser/token.cs]: changed here should be reverted.

File: src/Microsoft.WSMan.Management/ConfigProvider.cs [hunk 4]
```diff
@@ -2669,7 +2669,7 @@ private XmlDocument EnumerateResourceValue(object sessionobj, string ResourceURI
             {
                 try
                 {
-                    object value = ((IWSManSession)sessionobj).Enumerate(ResourceURI, "", "", 0);
+                    object value = ((IWSManSession)sessionobj).Enumerate(ResourceURI, string.Empty, "", 0);
```
Comment (daxian-dbw) [inline src/Microsoft.WSMan.Management/ConfigProvider.cs]: missing replacing an instance of `""`

File: src/Microsoft.WSMan.Management/ConfigProvider.cs [hunk 5]
```diff
@@ -2803,13 +2803,13 @@ private bool ContainResourceValue(object sessionobj, string ResourceURI, string
             {
                 if (ResourceURI.Contains(WSManStringLiterals.containerListener) || ResourceURI.Contains(WSManStringLiterals.containerPlugin) || ResourceURI.Contains(WSManStringLiterals.containerCertMapping))
                 {
-                    object value = ((IWSManSession)sessionobj).Enumerate(ResourceURI, "", "", 0);
+                    object value = ((IWSManSession)sessionobj).Enumerate(ResourceURI, string.Empty, "", 0);
```
Comment (daxian-dbw) [inline src/Microsoft.WSMan.Management/ConfigProvider.cs]: Here too.

**Comment by daxian-dbw** [comment]
@dantraMSFT I think the purpose is to enable the style checking about empty string (favor `string.Emtpy` over `""`). That rule cannot be turned on today due to too much noise.
The intern string should not be a concern, now it's one instance per assembly, but after replacing with `string.Empty` it will be one instance per process.

@iSazonov Thanks for spending efforts on driving the style consistency. I ran [`check-stringreplacementchange-ps1`](https://gist.github.com/daxian-dbw/3df8f9236a73a1d9b8c5cc913476d618#file-check-stringreplacementchange-ps1) with this PR and the results look good except for the three comments I left.

I quickly go through the changes and didn't notice any replacements in PowerShell scripts in `.cs` files, that's good because `string.Emtpy` will definitely fail in powershell script. However, since I went through so quickly, I could miss something, so please double check that you are not replacing anything in powershell script (script strings). After double checking that, the PR should be good.

File: src/System.Management.Automation/DscSupport/CimDSCParser.cs (hunk 1 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/DscSupport/CimDSCParser.cs]: As long as we don't need to change the string, I prefer constant. But that should be in a separate PR.

**Comment by iSazonov** [comment]
@dantraMSFT The PR has a purpose to turn CodeFactor into a more useful tool. Currently we see over 100000 issues most of which are style issues.

File: src/System.Management.Automation/engine/Modules/ScriptAnalysis.cs [hunk 6]
```diff
@@ -224,7 +224,7 @@ public override AstVisitAction VisitFunctionDefinition(FunctionDefinitionAst fun
         // as loose PS1 files)
         public override AstVisitAction VisitAssignmentStatement(AssignmentStatementAst assignmentStatementAst)
         {
-            // $env:PATH += "";$psScriptRoot""
+            // $env:PATH += string.Empty;$psScriptRoot""
```
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/Modules/ScriptAnalysis.cs]: Good catch!
Fixed.

File: src/Microsoft.WSMan.Management/ConfigProvider.cs (hunk 4 above)
Comment (iSazonov) [reply src/Microsoft.WSMan.Management/ConfigProvider.cs]: Fixed.

File: src/Microsoft.WSMan.Management/ConfigProvider.cs (hunk 5 above)
Comment (iSazonov) [reply src/Microsoft.WSMan.Management/ConfigProvider.cs]: Fixed.

**Review by daxian-dbw — APPROVED** [review]

**Pull request merged**